cross_connect_server/models/cross_connect_client.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2024 Akretion (http://www.akretion.com).
+# @author Florian Mounier <florian.mounier@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from datetime import datetime, timedelta, timezone
+from secrets import token_urlsafe
+
+import jwt
+
+from odoo import _, api, fields, models
+from odoo.exceptions import AccessDenied
+
+
+class CrossConnectClient(models.Model):
+    _name = "cross.connect.client"
+    _description = "Cross Connect Client"
+    _inherit = "server.env.mixin"
+
+    name = fields.Char(required=True)
+
+    endpoint_id = fields.Many2one(
+        "fastapi.endpoint",
+        required=True,
+        string="Endpoint",
+    )
+
+    api_key = fields.Char(
+        required=True,
+        string="API Key",
+        help="The API key to give to configure on the client.",
+        default=lambda self: self._generate_api_key(),
+    )
+
+    allowed_group_ids = fields.Many2many(
+        related="endpoint_id.cross_connect_allowed_group_ids",
+    )
+
+    group_ids = fields.Many2many(
+        "res.groups",
+        string="Groups",
+        help="The groups that this client belongs to.",
+        domain="[('id', 'in', allowed_group_ids)]",
+    )
+
+    user_ids = fields.One2many(
+        "res.users",
+        "cross_connect_client_id",
+        string="Users",
+        help="The users created by this cross connection.",
+    )
+    user_count = fields.Integer(
+        compute="_compute_user_count",
+        string="Cross Connected User Count",
+        help="The number of users created by this cross connection.",
+    )
+
+    @api.model
+    def _generate_api_key(self):
+        # generate random ~64 chars secret key
+        return token_urlsafe(64)
+
+    @api.depends("user_ids")
+    def _compute_user_count(self):
+        for record in self:
+            record.user_count = len(record.user_ids)
+
+    def _request_access(self, access_request):
+        # check groups
+        groups = self.env["res.groups"].browse(access_request.groups)
+        if groups - self.group_ids or not groups.exists():
+            raise AccessDenied(_("You are not allowed to access this endpoint."))
+
+        user = self.user_ids.filtered(
+            lambda u: u.cross_connect_client_user_id == access_request.id
+        )
+        vals = {
+            "login": access_request.login,
```

Comment:
As the login must be uniq in the server. We should customize the login

the following pattern seem to be optimal

```suggestion
            "login": f"{self.id}_{access_request.id}_{access_request.login}"
```

This will be rock solid
- 2 customers can have the same login without issue (self.id is uniq per customer)
- a customer can do nasty thing (true story!) like renaming a login and creating a new one with the previous login !